packages/core-app-api/src/apis/implementations/TranslationApi/I18nextTranslationApi.ts

```diff
@@ -0,0 +1,327 @@
+        i18n.addResourceBundle(
+          DEFAULT_LANGUAGE,
+          resource.id,
+          removeNulls(resource.messages),
+          true, // merge with existing translations
+          false, // do not overwrite translations
+        );
+      }
+    }
+
+    const instance = new I18nextTranslationApi(
+      i18n,
+      loader,
+      options.languageApi.getLanguage().language,
+    );
+
+    options.languageApi.language$().subscribe(({ language }) => {
+      instance.#changeLanguage(language);
+    });
+
+    return instance;
+  }
+
+  #i18n: I18n;
+  #loader: ResourceLoader;
+  #language: string;
+
+  /** Keep track of which refs we have registered default resources for */
+  #registeredRefs = new Set<string>();
+
+  private constructor(i18n: I18n, loader: ResourceLoader, language: string) {
+    this.#i18n = i18n;
+    this.#loader = loader;
+    this.#language = language;
+  }
+
+  getTranslation<TMessages extends { [key in string]: string }>(
+    translationRef: TranslationRef<string, TMessages>,
+  ): TranslationSnapshot<TMessages> {
+    const internalRef = toInternalTranslationRef(translationRef);
+
+    this.#registerDefaults(internalRef);
+
+    return this.#createSnapshot(internalRef);
+  }
+
+  translation$<TMessages extends { [key in string]: string }>(
+    translationRef: TranslationRef<string, TMessages>,
+  ): Observable<TranslationSnapshot<TMessages>> {
+    const internalRef = toInternalTranslationRef(translationRef);
+
+    this.#registerDefaults(internalRef);
+
+    return new ObservableImpl<TranslationSnapshot<TMessages>>(subscriber => {
+      let loadTicket = {}; // To check for stale loads
+      let lastSnapshotWasReady = false;
+
+      const loadResource = () => {
+        loadTicket = {};
+        const ticket = loadTicket;
+        this.#loader.load(this.#language, internalRef.id).then(
+          () => {
+            if (ticket === loadTicket) {
+              const snapshot = this.#createSnapshot(internalRef);
+              if (snapshot.ready || lastSnapshotWasReady) {
+                lastSnapshotWasReady = snapshot.ready;
+                subscriber.next(snapshot);
+              }
+            }
+          },
+          error => {
+            if (ticket === loadTicket) {
+              subscriber.error(Array.isArray(error) ? error[0] : error);
+            }
+          },
+        );
+      };
+
+      const onChange = () => {
+        const snapshot = this.#createSnapshot(internalRef);
+        if (lastSnapshotWasReady && !snapshot.ready) {
+          lastSnapshotWasReady = snapshot.ready;
+          subscriber.next(snapshot);
+        }
+
+        if (!snapshot.ready) {
+          loadResource();
+        }
+      };
+
+      const wasInitialized = this.#i18n.isInitialized;
+      if (!wasInitialized) {
+        this.#i18n.on('initialized', onChange);
+      }
+      this.#i18n.on('languageChanged', onChange);
```

Comment:
Is there any risk you think, that this can hit a "max listeners exceeded" error when there are very large numbers of translated elements present?

It might be interesting to have a test that makes a synthetic page with at least thousands of `t` on it that goes through the whole life cycle of initial state -> change to a language that takes some time to load -> done

Reply:
Was able to refactor this to get rid of the i18n listeners. We can set the i18next initialization to be immediate, and I switched to internal tracking of language change listeners.